File: existing-file/imperative/koan.go
package main

import (
	"errors"
	"fmt"
	"os"
)

func main() {
	path := os.Args[1]
	text, err := readFile(path)

	if err != nil {
		os.Exit(1)
	}

	fmt.Print(text)
}

func readFile(path string) ([]byte, error) {
	text, err := os.ReadFile(path)

	if err != nil {
		return nil, errors.New("Domain error")
	}

	return text, nil
}

File: existing-file/imperative/koan.py
import sys


def main():
    path = sys.argv[1]
    text = read_file_reliably(path)
    print(text)


def read_file_reliably(path):
    try:
        read_file(path)
    except FileNotFoundError as e:
        raise DomainError from e


def read_file(path):
    with open(path) as f:
        return f.read()


class DomainError(Exception):
    pass


if __name__ == '__main__':
    main()
